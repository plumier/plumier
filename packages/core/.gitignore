!src/generic-controller/filter-grammar.js