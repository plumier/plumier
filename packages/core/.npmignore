test
src
.npmignore
.gitignore
tsconfig.json
package-lock.json
tsconfig.build.json
yarn-error.log
yarn.lock