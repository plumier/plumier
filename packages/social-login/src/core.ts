import {
    ActionResult,
    bind,
    CustomMiddleware,
    DefaultFacility,
    HttpCookie,
    HttpStatusError,
    Invocation,
    invoke,
    PlumierApplication,
    response,
    RouteInfo,
    RouteMetadata,
    VirtualRoute,
    Class
} from "@plumier/core"
import Axios, { AxiosError } from "axios"
import crypto from "crypto"
import Csrf from "csrf"
import debug from "debug"
import { Context } from "koa"
import OAuth from "oauth-1.0a"
import qs from "querystring"
import { decorateMethod } from "tinspector"

// --------------------------------------------------------------------- //
// ------------------------------- TYPES ------------------------------- //
// --------------------------------------------------------------------- //

export enum CookieName {
    csrfSecret = "plum-oauth:csrf-secret",
    provider = "plum-oauth:provider"
}

export type SocialProvider = "Facebook" | "GitHub" | "Google" | "GitLab" | "Twitter"
export type OAuthVersion = "1.0a" | "2.0"
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type Optional<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;

export interface EndPoint {
    endpoint: string
    params: {}
}

export interface OAuthOptions {
    provider: SocialProvider
    clientId: string
    clientSecret: string
    token: EndPoint
    requestToken?: EndPoint
    profile: EndPoint & { transformer: (value: any) => Optional<OAuthUser, "provider"> }
    login: EndPoint
    oAuthVersion: OAuthVersion
}

export interface OAuthUser {
    provider: SocialProvider
    id: string,
    name: string,
    firstName: string,
    lastName: string,
    profilePicture: string,
    email?: string,
    gender?: string,
    dateOfBirth?: string
}

export interface OAuthProviderOption {
    clientId?: string,
    clientSecret?: string,
    loginEndPoint?: string,
    profileParams?: {}
}

interface ParserResult {
    user: Optional<OAuthUser, "provider">,
    profile: any,
    token: any
}

interface GeneratorResult {
    redirect: string,
    cookies: OAuthCookies
}

export type OAuth1Request = (opt: { url: string, method: "GET" | "POST", data: any }, token?: OAuth.Token) => Promise<any>

// --------------------------------------------------------------------- //
// ------------------------------- HELPER ------------------------------ //
// --------------------------------------------------------------------- //

const log = {
    debug: debug("@plumier/social-login:debug"),
    error: debug("@plumier/social-login:error")
}

export function splitName(name: string | undefined) {
    if (!name) return { firstName: "", lastName: "" }
    const idx = name.lastIndexOf(" ")
    if (idx === -1) return { firstName: name, lastName: "" }
    const lastName = name.substr(idx + 1)
    const firstName = name.substring(0, idx)
    return { firstName, lastName }
}

class OAuthCookies {
    static async oAuth2(provider: SocialProvider) {
        const csrf = new Csrf()
        const secret = await csrf.secret()
        return new OAuthCookies(secret, provider)
    }

    constructor(public secret: string, private provider: SocialProvider) { }

    toHttpCookies(): HttpCookie[] {
        return [
            { key: CookieName.csrfSecret, value: this.secret },
            { key: CookieName.provider, value: this.provider },
        ]
    }
}

export function oAuth1ARequest(apiKey: string, apiKeySecret: string): OAuth1Request {
    var oauth = new OAuth({
        consumer: {
            key: apiKey,
            secret: apiKeySecret
        },
        signature_method: 'HMAC-SHA1',
        hash_function: function (base_string, key) {
            return crypto.createHmac('sha1', key).update(base_string).digest('base64');
        }
    });
    return async ({ url, method, data }: { url: string, method: "GET" | "POST", data: any }, token?: OAuth.Token) => {
        const headers = oauth.toHeader(oauth.authorize({ url, method, data }, token))
        const resp = await Axios({ url, method, headers, data: method === "GET" ? undefined : data })
        return resp.data
    }
}

// --------------------------------------------------------------------- //
// ----------------------------- DECORATORS ---------------------------- //
// --------------------------------------------------------------------- //

export interface OAuthRedirectUriDecorator { name: "OAuthRedirectUriDecorator", provider: SocialProvider }

export function redirectUri(provider: SocialProvider | "General" = "General") {
    return decorateMethod(<OAuthRedirectUriDecorator>{ name: "OAuthRedirectUriDecorator", provider })
}

declare module "@plumier/core" {
    namespace bind {
        export function oAuthUser(): (target: any, name: string, index: number) => void
        export function oAuthToken(): (target: any, name: string, index: number) => void
        export function oAuthProfile(): (target: any, name: string, index: number) => void
    }
}

bind.oAuthUser = () => bind.custom(x => x.state.oAuthUser)
bind.oAuthToken = () => bind.custom(x => x.state.oAuthToken)
bind.oAuthProfile = () => bind.custom(x => x.state.oAuthProfile)



// --------------------------------------------------------------------- //
// ------------------------------ RESPONSE ----------------------------- //
// --------------------------------------------------------------------- // 

declare module "@plumier/core" {
    namespace response {
        function postMessage(message: any, origin?: string): ActionResult;
    }
}

response.postMessage = (message: any, origin?: string) => {
    return new ActionResult(`
    <!DOCTYPE html>
    <html>
        <title></title>
        <body>
            <div class="container"></div>
            <script type="text/javascript">
                var message = '${JSON.stringify(message)}';
                var origin ${!!origin ? `= '${origin}'` : ""};
                (function(){
                    window.onbeforeunload = function () {
                        window.opener.postMessage({ status: "Close" }, origin || window.location.origin);
                    };
                    window.opener.postMessage(JSON.parse(message), origin || window.location.origin);
                })()
            </script>
        </body>
    </html>
    `).setHeader("Content-Type", "text/html")
}

function clientRedirect(url: string) {
    // redirect client using JS to possibly set cookie before redirection
    return new ActionResult(`
    <!DOCTYPE html>
    <html>
        <title></title>
        <body>
            <div class="container"></div>
            <script type="text/javascript">
                const REDIRECT = '${url}';
                (function(){
                    window.location.href = REDIRECT;
                })()
            </script>
        </body>
    </html>
    `).setHeader("Content-Type", "text/html")
}

// --------------------------------------------------------------------- //
// ------------------------ LOGIN URL GENERATOR ------------------------ //
// --------------------------------------------------------------------- //

// oauth 2.0
async function oAuth20LoginGenerator(ctx: Context, option: OAuthOptions, redirect_uri: string): Promise<GeneratorResult> {
    const params: any = { ...option.login.params, ...ctx.query }
    params.redirect_uri = ctx.origin + redirect_uri
    params.client_id = option.clientId
    const cookies = await OAuthCookies.oAuth2(option.provider)
    const csrf = new Csrf()
    params.state = csrf.create(cookies.secret)
    const redirect = option.login.endpoint + "?" + qs.stringify(params)
    return { redirect, cookies }
}

// oauth 1.0a
async function oAuth10aLoginGenerator(ctx: Context, opt: OAuthOptions, redirectUri: string): Promise<GeneratorResult> {
    const request = oAuth1ARequest(opt.clientId, opt.clientSecret)
    const rawTokens = await request({
        url: opt.requestToken!.endpoint,
        method: "POST",
        data: { oauth_callback: ctx.origin + redirectUri }
    })
    const reqTokens: any = qs.parse(rawTokens)
    const params = { oauth_token: reqTokens.oauth_token, ...opt.login.params, ...ctx.query }
    const redirect = `${opt.login.endpoint}?${qs.stringify(params)}`
    const cookies = new OAuthCookies(reqTokens.oauth_token, opt.provider)
    return { redirect, cookies }
}


// --------------------------------------------------------------------- //
// --------------------------- PROFILE PARSER -------------------------- //
// --------------------------------------------------------------------- //

// oauth 2.0
async function oAuth20ProfileParser(ctx: Context, option: OAuthOptions, redirect_uri: string): Promise<ParserResult> {
    const req = ctx.request
    const secret = ctx.cookies.get(CookieName.csrfSecret)!
    if (!new Csrf().verify(secret, req.query.state)) {
        log.error("*** Invalid csrf token ***")
        log.error("CSRF Secret: %s", secret)
        log.error("CSRF Token: %s", req.query.state)
        throw new HttpStatusError(400)
    }
    const { clientId: client_id, clientSecret: client_secret } = option
    // request access token
    const { data: token } = await Axios.post<{ access_token: string }>(option.token.endpoint,
        { client_id, redirect_uri, client_secret, code: req.query.code, ...option.token.params },
        { headers: { Accept: 'application/json' } })
    // request profile
    const { data: profile } = await Axios.get(option.profile.endpoint, {
        params: option.profile.params,
        headers: { Authorization: `Bearer ${token.access_token}` }
    })
    const user = option.profile.transformer(profile)
    return { user, profile, token }
}

// oauth 1.0
async function oAuth10aProfileParser(ctx: Context, option: OAuthOptions, redirect_uri: string): Promise<ParserResult> {
    const req = ctx.request
    const secret = ctx.cookies.get(CookieName.csrfSecret)!
    if (secret !== req.query.oauth_token) {
        log.error("*** Invalid oauth_token ***")
        log.error("Saved token: %s", secret)
        log.error("Request token: %s", req.query.oauth_token)
        throw new HttpStatusError(400)
    }
    const request = oAuth1ARequest(option.clientId, option.clientSecret)
    const rawTokens = await request({
        url: option.token.endpoint, method: "POST",
        data: {
            oauth_token: req.query.oauth_token,
            oauth_verifier: req.query.oauth_verifier
        }
    })
    const token: any = qs.parse(rawTokens)
    const profile = await request({
        url: option.profile.endpoint, method: "GET",
        data: {}
    }, { key: token.oauth_token, secret: token.oauth_token_secret })
    const user = option.profile.transformer(profile)
    return { user, profile, token }
}

// --------------------------------------------------------------------- //
// ---------------------------- MIDDLEWARES ---------------------------- //
// --------------------------------------------------------------------- //

class OAuthLoginEndPointMiddleware implements CustomMiddleware {
    readonly generator = {
        "1.0a": oAuth10aLoginGenerator,
        "2.0": oAuth20LoginGenerator
    }

    constructor(private option: OAuthOptions, private loginPath: string, private redirectUri: string) { }

    async execute(invocation: Readonly<Invocation>): Promise<ActionResult> {
        if (invocation.ctx.path.toLowerCase() !== this.loginPath.toLowerCase()) return invocation.proceed()
        const { redirect, cookies } = await this.generator[this.option.oAuthVersion](invocation.ctx,
            this.option, this.redirectUri)
        return clientRedirect(redirect)
            .setCookie(cookies.toHttpCookies())
    }
}

class OAuthRedirectUriMiddleware implements CustomMiddleware {
    readonly parser = {
        "1.0a": oAuth10aProfileParser,
        "2.0": oAuth20ProfileParser
    }

    constructor(private option: OAuthOptions, private redirectUri: string) { }

    async execute(inv: Readonly<Invocation>): Promise<ActionResult> {
        const req = inv.ctx.request;
        if (inv.ctx.state.caller === "invoke") return inv.proceed()
        if (inv.ctx.path.toLocaleLowerCase() !== this.redirectUri.toLowerCase()) return inv.proceed()
        const provider = inv.ctx.cookies.get(CookieName.provider)
        if (provider !== this.option.provider) return inv.proceed()
        log.debug("Provider: %s", this.option.provider)
        const redirectUri = req.origin + req.path
        try {
            const result = await this.parser[this.option.oAuthVersion](inv.ctx, this.option, redirectUri)
            log.debug("OAuth User: %O", result.user)
            const provider = this.option.provider
            inv.ctx.state.oAuthUser = { ...result.user, provider }
            inv.ctx.state.oAuthToken = { ...result.token, provider }
            inv.ctx.state.oAuthProfile = { ...result.profile, provider }
        }
        catch (e) {
            log.error("*** OAuth Error ***", this.option.provider)
            log.error("Auth Code: %s", req.query.code)
            log.error("Redirect Uri: %s", redirectUri)
            if (e.response) {
                const error = (e as AxiosError)
                const response = error.response!;
                log.error("Request URL: %s %s", error.config.method, error.config.url)
                log.error("Request Data: %o", error.config.data)
                log.error("Request Headers: %o", error.config.headers)
                log.error("Error: %o", { status: response.status, message: response.statusText, data: response.data })
                throw new HttpStatusError(response.status, response.statusText)
            }
            throw e
        }
        return invoke(inv.ctx, inv.ctx.route!)
    }
}

// --------------------------------------------------------------------- //
// ------------------------------ FACILITY ----------------------------- //
// --------------------------------------------------------------------- //

export class OAuthProviderBaseFacility extends DefaultFacility {
    private option: Optional<OAuthOptions, "clientId" | "clientSecret">
    private loginEndpoint: string

    constructor(option: Optional<OAuthOptions, "clientId" | "clientSecret">, opt?: OAuthProviderOption) {
        super()
        this.option = {
            ...option,
            clientId: opt?.clientId,
            clientSecret: opt?.clientSecret,
        }
        if (opt && opt.profileParams)
            this.option.profile.params = { ...this.option.profile.params, ...opt.profileParams }
        this.loginEndpoint = opt?.loginEndPoint ?? `/auth/${this.option.provider.toLowerCase()}/login`
    }

    async generateRoutes(): Promise<VirtualRoute[]> {
        return [{
            kind: "VirtualRoute",
            method: "get",
            provider: this.constructor as Class,
            url: this.loginEndpoint,
            access: "Public",
            overridable: false,
            openApiOperation: {
                description: `Redirect request to ${this.option.provider} login page. This endpoint writes CSRF token on the client required to generate ${this.option.provider} login endpoint`,
                tags: ["Social Login"],
                parameters: [
                    {}
                ],
            }
        }]
    }

    getRedirectUri(routes: RouteMetadata[], provider: string) {
        return routes.find((x): x is RouteInfo => x.kind === "ActionRoute" && x.action.decorators
            .some((y: OAuthRedirectUriDecorator) => y.name === "OAuthRedirectUriDecorator" && y.provider === provider))
    }

    async initialize(app: Readonly<PlumierApplication>, routes: RouteMetadata[]) {
        const redirectUriRoute = this.getRedirectUri(routes, this.option.provider) ?? this.getRedirectUri(routes, "General")
        if (!redirectUriRoute) throw new Error(`OAuth: No ${this.option.provider} redirect uri handler found`)
        if (redirectUriRoute.url.search(":") > -1) throw new Error(`OAuth: Parameterized route is not supported on ${this.option.provider} callback uri`)
        if (redirectUriRoute.method !== "get") throw new Error(`OAuth: Redirect uri must have GET http method on ${redirectUriRoute.controller.name}.${redirectUriRoute.action.name}`)
        const clientIdKey = `PLUM_${this.option.provider.toUpperCase()}_CLIENT_ID`
        const clientSecretKey = `PLUM_${this.option.provider.toUpperCase()}_CLIENT_SECRET`
        const clientId = this.option.clientId ?? process.env[clientIdKey]
        if (!clientId) throw new Error(`OAuth: Client id for ${this.option.provider} not provided`)
        const clientSecret = this.option.clientSecret ?? process.env[clientSecretKey]
        if (!clientSecret) throw new Error(`OAuth: Client secret for ${this.option.provider} not provided`)
        const option = { ...this.option, clientId, clientSecret }
        app.use(new OAuthLoginEndPointMiddleware(option, this.loginEndpoint, redirectUriRoute.url))
        app.use(new OAuthRedirectUriMiddleware(option, redirectUriRoute.url))
    }
}