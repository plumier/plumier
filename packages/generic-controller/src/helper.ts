import { RelationDecorator, RelationPropertyDecorator } from "@plumier/core"
import reflect, { Class, generic, GenericTypeDecorator } from "@plumier/reflect"


// --------------------------------------------------------------------- //
// ----------------------- ENTITY RELATION HELPER ---------------------- //
// --------------------------------------------------------------------- //

const genericControllerRegistry = new Map<Class, boolean>()

function updateGenericControllerRegistry(cls: Class) {
    genericControllerRegistry.set(cls, true)
}

function getGenericControllerRelation(controller: Class) {
    const meta = reflect(controller)
    const types = generic.getGenericTypeParameters(controller)
    const parentEntityType = types[0]
    const entityType = types[1]
    const oneToMany = meta.decorators.find((x: RelationPropertyDecorator): x is RelationPropertyDecorator => x.kind === "plumier-meta:relation-prop-name")
    const relation = oneToMany!.name
    return { parentEntityType, entityType, relation, inverseProperty: oneToMany!.inverseProperty }
}

function getGenericControllerInverseProperty(controller: Class) {
    const rel = getGenericControllerRelation(controller)
    return rel.inverseProperty
}

function getGenericControllerOneToOneRelations(type: Class) {
    const meta = reflect(type)
    const result = []
    for (const prop of meta.properties) {
        if (prop.decorators.find((x: RelationDecorator) => x.kind === "plumier-meta:relation") && prop.typeClassification !== "Array") {
            result.push(prop)
        }
    }
    return result
}


export {
    getGenericControllerOneToOneRelations, getGenericControllerInverseProperty,
    getGenericControllerRelation, updateGenericControllerRegistry, genericControllerRegistry
}