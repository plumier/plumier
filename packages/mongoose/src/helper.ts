import { authorize } from "@plumier/core"
import reflect, { Class, useCache } from "@plumier/reflect"

import { RefDecorator } from "./types"

function normalizeEntityNoCache(type: Class) {
    const meta = reflect(type)
    reflect.flush(type)
    for (const prop of meta.properties) {
        const ref = prop.decorators.find((x: RefDecorator) => x.name === "MongooseRef")
        if (ref && prop.typeClassification === "Array") {
            const decorators = [
                authorize.readonly(),
                authorize.writeonly()
            ]
            Reflect.decorate(decorators, type.prototype, prop.name, void 0)
        }
    }
    return { success: true }
}

const normalizeEntityCache = new Map<Class, any>()

export const normalizeEntity = useCache(normalizeEntityCache, normalizeEntityNoCache, x => x)