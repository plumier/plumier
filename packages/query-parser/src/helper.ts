import { AuthorizeDecorator, AuthorizerContext, authPolicy, AuthPolicy, EntityAuthPolicy, executeAuthorizer } from "@plumier/core"
import { Class, generic } from "@plumier/reflect"


const ParserAst = Symbol()

function getDecoratorType(controller: Class, expType: string | Class) {
    // extract generic type from controller if string type provided
    return typeof expType === "string" ? generic.getGenericType(controller, expType) as Class : expType
}

async function isAuthorized(decorators: AuthorizeDecorator[], ctx: AuthorizerContext): Promise<boolean> {
    const authPolicies = ctx.ctx.config.authPolicies.map(x => new x())
    const policies: { decorator: AuthorizeDecorator, isEntityPolicy: boolean }[] = []
    const staticPolicies = []
    // flatten decorator to make each decorator only have one policy
    // and add flag if its an entity policy
    // to make them easily filtered by its flag
    for (const decorator of decorators) {
        for (const policy of decorator.policies) {
            const isEntityPolicy = authPolicies.some(x => x.name === policy && x instanceof EntityAuthPolicy)
            const dec: AuthorizeDecorator = { ...decorator, policies: [policy] }
            if (!isEntityPolicy)
                staticPolicies.push(dec)
            policies.push({ decorator: dec, isEntityPolicy })
        }
    }
    if (policies.every(x => x.isEntityPolicy)) return false
    return executeAuthorizer(staticPolicies, ctx)
}

export { getDecoratorType, ParserAst, isAuthorized }