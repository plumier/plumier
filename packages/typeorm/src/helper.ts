import { authorize, Class, entity } from "@plumier/core"
import reflect, { noop, useCache } from "@plumier/reflect"
import { getMetadataArgsStorage } from "typeorm"
import { parse } from "acorn"

function inverseSideParser(expr: ((t: any) => any)) {
    const node = parse(expr.toString(), { ecmaVersion: 2020 })
    return getMemberExpression(node)
}

function getContent(node: any): any {
    switch (node.type) {
        case "Program":
        case "BlockStatement":
            return node.body[node.body.length - 1]
        case "ArrowFunctionExpression":
            return node.body
        case "ExpressionStatement":
            return node.expression
        case "ReturnStatement":
            return node.argument
    }
}

function getMemberExpression(node: any): string {
    const content = getContent(node)
    if (content.type === "MemberExpression")
        return content.property.name
    else
        return getMemberExpression(content)
}


function normalizeEntityNoCache(type: Class) {
    const parent: Class = Object.getPrototypeOf(type)
    // loop through parent entities 
    if (!!parent.prototype) normalizeEntity(parent)
    const storage = getMetadataArgsStorage();
    const columns = storage.filterColumns(type)
    for (const col of columns) {
        Reflect.decorate([noop()], (col.target as Function).prototype, col.propertyName, void 0)
        if (col.options.primary)
            Reflect.decorate([entity.primaryId(), authorize.readonly()], (col.target as Function).prototype, col.propertyName, void 0)
    }
    const relations = storage.filterRelations(type)
    for (const col of relations) {
        const rawType: Class = (col as any).type()
        if (col.relationType === "many-to-many" || col.relationType === "one-to-many") {
            const inverseProperty = inverseSideParser(col.inverseSideProperty as any)
            const decorators = [
                reflect.type(x => [rawType]),
                entity.relation({ inverseProperty }),
                authorize.readonly(),
                authorize.writeonly()
            ]
            Reflect.decorate(decorators, (col.target as Function).prototype, col.propertyName, void 0)
        }
        else {
            Reflect.decorate([reflect.type(x => rawType), entity.relation()], (col.target as Function).prototype, col.propertyName, void 0)
        }
    }
}

const normalizeEntityCache = new Map<Class, any>()

export const normalizeEntity = useCache(normalizeEntityCache, normalizeEntityNoCache, x => x)