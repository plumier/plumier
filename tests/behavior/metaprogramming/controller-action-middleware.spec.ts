import { middleware } from "@plumier/core"
import supertest from "supertest"

import { fixture } from "../helper"


describe("Action Middleware", () => {
    it("Should able to access metadata from action controller", async () => {
        const fn = jest.fn()
        @middleware.use(x => {
            fn(x.metadata)
            return x.proceed()
        })
        class AnimalController {
            get(id: string) { return { id } }
        }
        const app = await fixture(AnimalController).initialize()
        await supertest(app.callback())
            .get("/animal/get?id=1234")
            .expect(200, { id: "1234" })
        expect(fn.mock.calls).toMatchSnapshot()
    })

    it("Should able to get parameter by name", async () => {
        const fn = jest.fn()
        @middleware.use(x => {
            fn(x.metadata.actionParams.get<string>("id"))
            return x.proceed()
        })
        class AnimalController {
            get(id: string) { return { id } }
        }
        const app = await fixture(AnimalController).initialize()
        await supertest(app.callback())
            .get("/animal/get?id=1234")
            .expect(200, { id: "1234" })
        expect(fn.mock.calls).toMatchSnapshot()
    })

    it("Should return undefined if provided wrong parameter name", async () => {
        const fn = jest.fn()
        @middleware.use(x => {
            fn(x.metadata.actionParams.get("UserId"))
            return x.proceed()
        })
        class AnimalController {
            get(id: string) { return { id } }
        }
        const app = await fixture(AnimalController).initialize()
        await supertest(app.callback())
            .get("/animal/get?id=1234")
            .expect(200, { id: "1234" })
        expect(fn.mock.calls).toMatchSnapshot()
    })

    it("Should able to get parameter by number", async () => {
        const fn = jest.fn()
        @middleware.use(x => {
            fn(x.metadata.actionParams.get(0))
            return x.proceed()
        })
        class AnimalController {
            get(id: string) { return { id } }
        }
        const app = await fixture(AnimalController).initialize()
        await supertest(app.callback())
            .get("/animal/get?id=1234")
            .expect(200, { id: "1234" })
        expect(fn.mock.calls).toMatchSnapshot()
    })

    it("Should able to get check if has parameter", async () => {
        const fn = jest.fn()
        @middleware.use(x => {
            fn(x.metadata.actionParams.hasName("ID"))
            fn(x.metadata.actionParams.hasName("UserID"))
            return x.proceed()
        })
        class AnimalController {
            get(id: string) { return { id } }
        }
        const app = await fixture(AnimalController).initialize()
        await supertest(app.callback())
            .get("/animal/get?id=1234")
            .expect(200, { id: "1234" })
        expect(fn.mock.calls).toMatchSnapshot()
    })

    it("Should able to get all parameter names", async () => {
        const fn = jest.fn()
        @middleware.use(x => {
            fn(x.metadata.actionParams.names())
            return x.proceed()
        })
        class AnimalController {
            get(id: string, name: string) { return { id, name } }
        }
        const app = await fixture(AnimalController).initialize()
        await supertest(app.callback())
            .get("/animal/get?id=1234&name=mimi")
            .expect(200, { id: "1234", name: "mimi" })
        expect(fn.mock.calls).toMatchSnapshot()
    })

    it("Should able to get all parameter values", async () => {
        const fn = jest.fn()
        @middleware.use(x => {
            fn(x.metadata.actionParams.values())
            return x.proceed()
        })
        class AnimalController {
            get(id: string, name: string) { return { id, name } }
        }
        const app = await fixture(AnimalController).initialize()
        await supertest(app.callback())
            .get("/animal/get?id=1234&name=mimi")
            .expect(200, { id: "1234", name: "mimi" })
        expect(fn.mock.calls).toMatchSnapshot()
    })

    it("Should able spread invocation object", async () => {
        const fn = jest.fn()
        @middleware.use(({ ctx, metadata, proceed }) => {
            fn(metadata?.controller.name)
            fn(ctx.url)
            return proceed()
        })
        class AnimalController {
            get(id: string) { return { id } }
        }
        const app = await fixture(AnimalController).initialize()
        await supertest(app.callback())
            .get("/animal/get?id=1234")
            .expect(200, { id: "1234" })
        expect(fn.mock.calls).toMatchSnapshot()
    })

    it("Should able to get current metadata info when applied on controller", async () => {
        const fn = jest.fn()
        @middleware.use((invocation) => {
            fn(invocation.metadata.current)
            return invocation.proceed()
        })
        class AnimalController {
            get(id: string) { return { id } }
        }
        const app = await fixture(AnimalController).initialize()
        await supertest(app.callback())
            .get("/animal/get?id=1234")
            .expect(200, { id: "1234" })
        expect(fn.mock.calls).toMatchSnapshot()
    })

    it("Should able to get current metadata info when applied on method", async () => {
        const fn = jest.fn()
        class AnimalController {
            @middleware.use((invocation) => {
                fn(invocation.metadata.current)
                return invocation.proceed()
            })
            get(id: string) { return { id } }
        }
        const app = await fixture(AnimalController).initialize()
        await supertest(app.callback())
            .get("/animal/get?id=1234")
            .expect(200, { id: "1234" })
        expect(fn.mock.calls).toMatchSnapshot()
    })
})
