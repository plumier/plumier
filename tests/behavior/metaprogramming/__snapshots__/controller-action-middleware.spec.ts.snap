// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action Middleware Should able spread invocation object 1`] = `
Array [
  Array [
    "AnimalController",
  ],
  Array [
    "/animal/get?id=1234",
  ],
]
`;

exports[`Action Middleware Should able to access metadata from action controller 1`] = `
Array [
  Array [
    MetadataImpl {
      "access": undefined,
      "action": Object {
        "decorators": Array [],
        "kind": "Method",
        "name": "get",
        "owner": Array [
          AnimalController,
        ],
        "parameters": Array [
          Object {
            "decorators": Array [],
            "fields": Object {},
            "index": 0,
            "kind": "Parameter",
            "name": "id",
            "owner": Array [
              AnimalController,
            ],
            "type": undefined,
            "typeClassification": undefined,
          },
        ],
        "returnType": undefined,
        "typeClassification": undefined,
      },
      "actionParams": ParameterMetadata {
        "meta": Array [
          Object {
            "decorators": Array [],
            "fields": Object {},
            "index": 0,
            "kind": "Parameter",
            "name": "id",
            "owner": Array [
              AnimalController,
            ],
            "type": undefined,
            "typeClassification": undefined,
          },
        ],
        "parameters": Array [
          "1234",
        ],
      },
      "controller": Object {
        "ctor": Object {
          "kind": "Constructor",
          "name": "constructor",
          "parameters": Array [],
        },
        "decorators": Array [
          Object {
            "name": "Middleware",
            "target": "Controller",
            "value": Array [
              ,
            ],
            Symbol(tinspector:decoratorOption): Object {
              "allowMultiple": true,
              "inherit": true,
            },
          },
        ],
        "kind": "Class",
        "methods": Array [
          Object {
            "decorators": Array [],
            "kind": "Method",
            "name": "get",
            "owner": Array [
              AnimalController,
            ],
            "parameters": Array [
              Object {
                "decorators": Array [],
                "fields": Object {},
                "index": 0,
                "kind": "Parameter",
                "name": "id",
                "owner": Array [
                  AnimalController,
                ],
                "type": undefined,
                "typeClassification": undefined,
              },
            ],
            "returnType": undefined,
            "typeClassification": undefined,
          },
        ],
        "name": "AnimalController",
        "owner": Array [
          Object,
          AnimalController,
        ],
        "properties": Array [],
        "super": Object,
        "type": AnimalController,
        "typeClassification": "Class",
      },
      "current": Object {
        "ctor": Object {
          "kind": "Constructor",
          "name": "constructor",
          "parameters": Array [],
        },
        "decorators": Array [
          Object {
            "name": "Middleware",
            "target": "Controller",
            "value": Array [
              ,
            ],
            Symbol(tinspector:decoratorOption): Object {
              "allowMultiple": true,
              "inherit": true,
            },
          },
        ],
        "kind": "Class",
        "methods": Array [
          Object {
            "decorators": Array [],
            "kind": "Method",
            "name": "get",
            "owner": Array [
              AnimalController,
            ],
            "parameters": Array [
              Object {
                "decorators": Array [],
                "fields": Object {},
                "index": 0,
                "kind": "Parameter",
                "name": "id",
                "owner": Array [
                  AnimalController,
                ],
                "type": undefined,
                "typeClassification": undefined,
              },
            ],
            "returnType": undefined,
            "typeClassification": undefined,
          },
        ],
        "name": "AnimalController",
        "owner": Array [
          Object,
          AnimalController,
        ],
        "properties": Array [],
        "super": Object,
        "type": AnimalController,
        "typeClassification": "Class",
      },
    },
  ],
]
`;

exports[`Action Middleware Should able to get all parameter names 1`] = `
Array [
  Array [
    Array [
      "id",
      "name",
    ],
  ],
]
`;

exports[`Action Middleware Should able to get all parameter values 1`] = `
Array [
  Array [
    Array [
      "1234",
      "mimi",
    ],
  ],
]
`;

exports[`Action Middleware Should able to get check if has parameter 1`] = `
Array [
  Array [
    true,
  ],
  Array [
    false,
  ],
]
`;

exports[`Action Middleware Should able to get current metadata info when applied on controller 1`] = `
Array [
  Array [
    Object {
      "ctor": Object {
        "kind": "Constructor",
        "name": "constructor",
        "parameters": Array [],
      },
      "decorators": Array [
        Object {
          "name": "Middleware",
          "target": "Controller",
          "value": Array [
            ,
          ],
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": true,
            "inherit": true,
          },
        },
      ],
      "kind": "Class",
      "methods": Array [
        Object {
          "decorators": Array [],
          "kind": "Method",
          "name": "get",
          "owner": Array [
            AnimalController,
          ],
          "parameters": Array [
            Object {
              "decorators": Array [],
              "fields": Object {},
              "index": 0,
              "kind": "Parameter",
              "name": "id",
              "owner": Array [
                AnimalController,
              ],
              "type": undefined,
              "typeClassification": undefined,
            },
          ],
          "returnType": undefined,
          "typeClassification": undefined,
        },
      ],
      "name": "AnimalController",
      "owner": Array [
        Object,
        AnimalController,
      ],
      "properties": Array [],
      "super": Object,
      "type": AnimalController,
      "typeClassification": "Class",
    },
  ],
]
`;

exports[`Action Middleware Should able to get current metadata info when applied on method 1`] = `
Array [
  Array [
    Object {
      "decorators": Array [
        Object {
          "name": "Middleware",
          "target": "Action",
          "value": Array [
            ,
          ],
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": true,
            "inherit": true,
          },
        },
      ],
      "kind": "Method",
      "name": "get",
      "owner": Array [
        AnimalController,
      ],
      "parameters": Array [
        Object {
          "decorators": Array [],
          "fields": Object {},
          "index": 0,
          "kind": "Parameter",
          "name": "id",
          "owner": Array [
            AnimalController,
          ],
          "type": String,
          "typeClassification": "Primitive",
        },
      ],
      "returnType": undefined,
      "typeClassification": undefined,
    },
  ],
]
`;

exports[`Action Middleware Should able to get parameter by name 1`] = `
Array [
  Array [
    "1234",
  ],
]
`;

exports[`Action Middleware Should able to get parameter by number 1`] = `
Array [
  Array [
    "1234",
  ],
]
`;

exports[`Action Middleware Should return undefined if provided wrong parameter name 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;
